pkg/p2p/libp2p/internal/handshake/handshake.go

```diff
@@ -261,47 +284,55 @@ func (s *Service) Handle(ctx context.Context, stream p2p.Stream, remoteMultiaddr
 
 	welcomeMessage := s.GetWelcomeMessage()
 
-	if err := w.WriteMsgWithContext(ctx, &pb.SynAck{
-		Syn: &pb.Syn{
-			ObservedUnderlay: fullRemoteMABytes,
-		},
-		Ack: &pb.Ack{
-			Address: &pb.BzzAddress{
-				Underlay:  advertisableUnderlayBytes,
-				Overlay:   bzzAddress.Overlay.Bytes(),
-				Signature: bzzAddress.Signature,
+	if err := w.WriteMsgWithContext(ctx, &pb.Handshake{
+		Payload: &pb.Handshake_SynAck{
+			SynAck: &pb.HandshakeSynAck{
+				ObservedUnderlay: fullRemoteMABytes,
+				Address: &pb.BzzAddress{
+					Underlay:  advertisableUnderlayBytes,
+					Overlay:   bzzAddress.Overlay.Bytes(),
+					Signature: bzzAddress.Signature,
+				},
+				NetworkID: s.networkID,
+				Capabilities: &pb.Capabilities{
+					FullNode:     s.fullNode,
+					TraceHeaders: s.traceHeaders,
+				},
+				Nonce:          s.nonce,
+				WelcomeMessage: welcomeMessage,
 			},
-			NetworkID:      s.networkID,
-			FullNode:       s.fullNode,
-			Nonce:          s.nonce,
-			WelcomeMessage: welcomeMessage,
 		},
 	}); err != nil {
 		s.metrics.SynAckTxFailed.Inc()
 		return nil, fmt.Errorf("write synack message: %w", err)
 	}
 	s.metrics.SynAckTx.Inc()
 
-	var ack pb.Ack
-	if err := r.ReadMsgWithContext(ctx, &ack); err != nil {
+	var ackMsg pb.Handshake
+	if err := r.ReadMsgWithContext(ctx, &ackMsg); err != nil {
 		s.metrics.AckRxFailed.Inc()
 		return nil, fmt.Errorf("read ack message: %w", err)
 	}
 	s.metrics.AckRx.Inc()
 
+	ack := ackMsg.GetAck()
+	if ack == nil {
+		return nil, fmt.Errorf("expected ack message")
+	}
+
 	if ack.NetworkID != s.networkID {
 		return nil, ErrNetworkIDIncompatible
 	}
 
 	overlay := swarm.NewAddress(ack.Address.Overlay)
 
 	if s.picker != nil {
-		if !s.picker.Pick(p2p.Peer{Address: overlay, FullNode: ack.FullNode}) {
+		if !s.picker.Pick(p2p.Peer{Address: overlay, FullNode: ack.Capabilities.FullNode}) {
```

Comment:
Protocol Buffers does not automatically create empty message objects for missing fields and it can cause nil pointer dereference here since there is no nil check on Capabilities.